File: src/lib/terminal.js
'use strict';

const pty = require('node-pty');
const { v4: uuidv4 } = require('uuid');
const os = require('os');

class TerminalManager {
  constructor(io) {
    this.io = io;
    this.sessions = new Map();
    this.GRACE_PERIOD = 5 * 60 * 1000; // 5 mins
    this.MAX_BUFFER = 128 * 1024;
    this.TRIM_SIZE = 64 * 1024;
  }

  create(socket, data = {}) {
    const id = data.id || uuidv4();
    const shell = process.env.SHELL || '/bin/bash';
    const cols = data.cols || 80;
    const rows = data.rows || 24;

    try {
      const ptyProc = pty.spawn(shell, [], {
        name: 'xterm-256color',
        cols, rows,
        cwd: data.cwd || os.homedir(),
        env: { ...process.env, TERM: 'xterm-256color', COLORTERM: 'truecolor' }
      });

      const session = {
        pty: ptyProc,
        socketId: socket.id,
        buffer: '',
        killTimer: null,
        dataListener: null
      };

      this.sessions.set(id, session);
      this.attach(id, socket);

      ptyProc.onExit(({ exitCode }) => {
        socket.emit('terminal:exit', { id, exitCode });
        this.sessions.delete(id);
      });

      socket.emit('terminal:created', { id });
      return id;
    } catch (err) {
      socket.emit('terminal:error', { id, error: err.message });
      return null;
    }
  }

  attach(id, socket) {
    const session = this.sessions.get(id);
    if (!session) return;

    if (session.dataListener) {
      try { session.dataListener.dispose(); } catch {}
    }

    session.socketId = socket.id;
    session.dataListener = session.pty.onData((chunk) => {
      session.buffer += chunk;
      if (session.buffer.length > this.MAX_BUFFER) {
        session.buffer = session.buffer.slice(-this.TRIM_SIZE);
      }
      socket.emit('terminal:data', { id, data: chunk });
    });
  }

  reconnect(socket, { id, cols, rows }) {
    const session = this.sessions.get(id);
    if (!session) {
      socket.emit('terminal:reconnect:fail', { id });
      return;
    }

    if (session.killTimer) {
      clearTimeout(session.killTimer);
      session.killTimer = null;
    }

    if (cols && rows) {
      try { session.pty.resize(cols, rows); } catch {}
    }

    this.attach(id, socket);
    socket.emit('terminal:replay', { id, buffer: session.buffer });
  }

  write(id, data) {
    const s = this.sessions.get(id);
    if (s) { try { s.pty.write(data); } catch {} }
  }

  resize(id, cols, rows) {
    const s = this.sessions.get(id);
    if (s) { try { s.pty.resize(cols, rows); } catch {} }
  }

  kill(id) {
    const s = this.sessions.get(id);
    if (s) {
      if (s.killTimer) clearTimeout(s.killTimer);
      try { s.pty.kill(); } catch {}
      this.sessions.delete(id);
    }
  }

  handleDisconnect(socketId) {
    for (const [id, s] of this.sessions.entries()) {
      if (s.socketId === socketId) {
        s.socketId = null;
        s.killTimer = setTimeout(() => {
          const ss = this.sessions.get(id);
          if (ss && !ss.socketId) {
            this.kill(id);
          }
        }, this.GRACE_PERIOD);
      }
    }
  }
}

module.exports = TerminalManager;

File: pkg/terminal/terminal.go
package terminal

import (
	"os"
	"os/exec"
	"sync"

	"github.com/creack/pty"
)

type Session struct {
	ID        string
	Pty       *os.File
	Cmd       *exec.Cmd
	Buffer    []byte
	BufferMu  sync.Mutex
	listeners map[chan []byte]bool
	mu        sync.Mutex
}

type Manager struct {
	sessions map[string]*Session
	mu       sync.Mutex
}

func NewManager() *Manager {
	return &Manager{
		sessions: make(map[string]*Session),
	}
}

func (m *Manager) Create(id, cwd string, cols, rows uint16) (*Session, error) {
	shell := os.Getenv("SHELL")
	if shell == "" {
		shell = "/bin/bash"
	}

	// -i for interactive mode to enable job control
	cmd := exec.Command(shell, "-i")
	if cwd != "" {
		cmd.Dir = cwd
	}
	cmd.Env = append(os.Environ(), "TERM=xterm-256color")

	f, err := pty.StartWithAttrs(cmd, &pty.Winsize{Cols: cols, Rows: rows}, nil)
	if err != nil {
		return nil, err
	}

	session := &Session{
		ID:        id,
		Pty:       f,
		Cmd:       cmd,
		listeners: make(map[chan []byte]bool),
	}

	m.mu.Lock()
	m.sessions[id] = session
	m.mu.Unlock()

	// Handle exit
	go func() {
		cmd.Wait()
		m.mu.Lock()
		delete(m.sessions, id)
		m.mu.Unlock()
		f.Close()
	}()

	// Start central reader
	go session.readLoop()

	return session, nil
}

func (s *Session) readLoop() {
	buf := make([]byte, 4096)
	for {
		n, err := s.Pty.Read(buf)
		if n > 0 {
			data := make([]byte, n)
			copy(data, buf[:n])

			// Update buffer
			s.BufferMu.Lock()
			s.Buffer = append(s.Buffer, data...)
			if len(s.Buffer) > 128*1024 {
				s.Buffer = s.Buffer[len(s.Buffer)-64*1024:]
			}
			s.BufferMu.Unlock()

			// Broadcast
			s.mu.Lock()
			for l := range s.listeners {
				select {
				case l <- data:
				default:
					// Drop if listener is too slow
				}
			}
			s.mu.Unlock()
		}
		if err != nil {
			break
		}
	}
}

func (s *Session) Attach() chan []byte {
	s.mu.Lock()
	defer s.mu.Unlock()
	c := make(chan []byte, 100)
	s.listeners[c] = true
	return c
}

func (s *Session) Detach(c chan []byte) {
	s.mu.Lock()
	defer s.mu.Unlock()
	delete(s.listeners, c)
}

func (m *Manager) Get(id string) *Session {
	m.mu.Lock()
	defer m.mu.Unlock()
	return m.sessions[id]
}

func (s *Session) Write(data []byte) (int, error) {
	return s.Pty.Write(data)
}

func (s *Session) Resize(cols, rows uint16) error {
	return pty.Setsize(s.Pty, &pty.Winsize{Cols: cols, Rows: rows})
}

func (s *Session) Kill() error {
	if s.Cmd.Process != nil {
		return s.Cmd.Process.Kill()
	}
	return nil
}
